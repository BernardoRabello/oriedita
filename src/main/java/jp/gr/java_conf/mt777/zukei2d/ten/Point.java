package jp.gr.java_conf.mt777.zukei2d.ten;

public class Point {
    //Used to represent point coordinates, direction vectors, etc.
    double x, y;

    public Point(Point p) {
        set(p);
    }

    public Point() {
        x = 0;
        y = 0;
    }

    public Point(double i, double j) {
        x = i;
        y = j;
    }

    public Point(double a, Point p, double b, Point q) {
        x = a * p.getX() + b * q.getX();
        y = a * p.getY() + b * q.getY();
    }

    public void display(String str0) {
        System.out.println(str0 + " (" + x + " , " + y + ")");
    }

    public void set(Point p) {
        x = p.getX();
        y = p.getY();
    }

    public void set(double i, double j) {
        x = i;
        y = j;
    }

    public void set(double a, Point p, double b, Point q) {
        x = a * p.getX() + b * q.getX();
        y = a * p.getY() + b * q.getY();
    }

    public void setX(double xx) {
        x = xx;
    }

    public void setY(double yy) {
        y = yy;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public void reset() {
        x = 0.0;
        y = 0.0;
    }

    public void parallel_move(double x1, double y1) {
        x = x + x1;
        y = y + y1;
    }

    /**
     * Function to find the distance (double) to other points ----------------------------------------------------
     */
    public double distance(Point p) {
        double x1 = p.getX() - x, y1 = p.getY() - y;
        return Math.sqrt(x1 * x1 + y1 * y1);
    }

    /**
     * A function that finds the square of the distance to another point ----------------------------------------------------
     */
    public double distanceSquared(Point p) {
        double x1 = p.getX() - x, y1 = p.getY() - y;
        return x1 * x1 + y1 * y1;
    }

    /**
     * When looking at the own Point as a reference, the positions of other points are returned as Points.
     */
    public Point other_Point_position(Point taPoint) {
        Point rPoint = new Point();
        rPoint.setX(taPoint.getX() - x);
        rPoint.setY(taPoint.getY() - y);
        return rPoint;
    }

    public void move(Point addPoint) {
        x = x + addPoint.getX();
        y = y + addPoint.getY();
    }
}
