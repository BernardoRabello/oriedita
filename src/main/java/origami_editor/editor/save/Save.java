package origami_editor.editor.save;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import origami.crease_pattern.element.Circle;
import origami.crease_pattern.element.LineSegment;
import origami.crease_pattern.element.Point;
import origami.data.save.LineSegmentSave;
import origami.data.save.PointSave;
import origami_editor.editor.databinding.ApplicationModel;
import origami_editor.editor.databinding.CanvasModel;
import origami_editor.editor.databinding.FoldedFigureModel;
import origami_editor.editor.databinding.GridModel;
import origami_editor.editor.drawing.tools.Camera;

import java.io.Serializable;

import static com.fasterxml.jackson.annotation.JsonSubTypes.Type;
import static com.fasterxml.jackson.annotation.JsonTypeInfo.Id;

@JsonTypeInfo(use = Id.NAME, property = "@version", defaultImpl = SaveV1.class)
@JsonSubTypes({@Type(SaveV1.class)})
public interface Save extends PointSave, LineSegmentSave, Serializable {
    ApplicationModel getApplicationModel();
    void setApplicationModel(ApplicationModel applicationModel);

    Camera getCreasePatternCamera();
    void setCreasePatternCamera(Camera creasePatternCamera);

    CanvasModel getCanvasModel();
    void setCanvasModel(CanvasModel canvasModel);

    GridModel getGridModel();
    void setGridModel(GridModel gridModel);

    FoldedFigureModel getFoldedFigureModel();
    void setFoldedFigureModel(FoldedFigureModel foldedFigureModel);

    void set(Save save);
    void add(Save save);

    /**
     * Returns if this save contains lines which are not savable to a .cp file without losing information.
     */
    boolean canSaveAsCp();
}
