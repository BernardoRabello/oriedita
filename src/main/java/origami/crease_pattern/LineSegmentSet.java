package origami.crease_pattern;

import origami.crease_pattern.element.LineColor;
import origami.crease_pattern.element.LineSegment;
import origami.crease_pattern.element.Point;
import origami_editor.editor.Save;

import java.util.ArrayList;
import java.util.List;

/**
 * Collection of line segments. Used when calculating folded shapes and two-colored cp.
 */
public class LineSegmentSet {
    List<LineSegment> lineSegments = new ArrayList<>(); //Instantiation of line segments

    public LineSegmentSet() {
        reset();
    }

    public LineSegmentSet(PointSet pointSet) {
        reset(pointSet.getNumLines());
        for (int i = 1; i <= pointSet.getNumLines(); i++) {
            lineSegments.get(i-1).set(pointSet.getPoint(pointSet.getBegin(i)), pointSet.getPoint(pointSet.getEnd(i)), pointSet.getColor(i), LineSegment.ActiveState.INACTIVE_0);
        }
    }

    public void reset() {
        lineSegments.clear();
    }

    public void set(LineSegmentSet ss) {
        reset(ss.getNumLineSegments());

        for (int i = 0; i < lineSegments.size(); i++) {
            LineSegment s = lineSegments.get(i);
            s.set(ss.lineSegments.get(i));
        }
    }

    //Get the total number of line segments
    public int getNumLineSegments() {
        return lineSegments.size();
    }

    public void reset(int size) {
        reset();

        for (int i = 0; i < size; i++) {
            lineSegments.add(new LineSegment());
        }
    }

    //Get the endpoint of the i-th line segment
    public Point getA(int i) {
        LineSegment s = lineSegments.get(i);
        return s.getA();
    }

    public Point getB(int i) {
        LineSegment s = lineSegments.get(i);
        return s.getB();
    }

    /**
     * Output the color of the i-th line segment
     */
    public LineColor getColor(int i) {
        LineSegment s = lineSegments.get(i);
        return s.getColor();
    }

    public void setSave(Save memo1) {
        lineSegments.clear();
        for (LineSegment s :
                memo1.getLineSegments()) {
            LineSegment s0 = new LineSegment();
            s0.set(s);
            lineSegments.add(s0);
        }
    }

    //Remove dotted line segments
    public void point_removal() {
        lineSegments.removeIf(s -> OritaCalc.equal(s.getA(), s.getB()));
    }

    /**
     * When there are two completely overlapping line segments, the one with the latest number is deleted.
     */
    public void overlapping_line_removal(double r) {
        List<Boolean> removal_flg = new ArrayList<>();
        List<LineSegment> snew = new ArrayList<>();
        for (int i = 0; i < lineSegments.size(); i++) {
            removal_flg.add(false);
            snew.add(new LineSegment());
        }

        for (int i = 0; i < lineSegments.size(); i++) {
            LineSegment si = lineSegments.get(i);
            for (int j = i + 1; j < lineSegments.size(); j++) {
                LineSegment sj = lineSegments.get(j);
                if (r <= -9999.9) {
                    if (OritaCalc.determineLineSegmentIntersection(si, sj) == LineSegment.Intersection.PARALLEL_EQUAL_31) {
                        removal_flg.set(j, true);
                    }
                } else {
                    if (OritaCalc.determineLineSegmentIntersection(si, sj, r, r) == LineSegment.Intersection.PARALLEL_EQUAL_31) {
                        removal_flg.set(j, true);
                    }
                }
            }
        }

        int smax = 0;
        for (int i = 0; i < lineSegments.size(); i++) {
            if (!removal_flg.get(i)) {
                LineSegment si = lineSegments.get(i);
                snew.get(smax).set(si);
                smax = smax + 1;
            }
        }

        if (smax > lineSegments.size()) {
            while (smax > lineSegments.size()) {
                lineSegments.add(new LineSegment());
            }
        }
        for (int i = 0; i < lineSegments.size(); i++) {
            LineSegment si = lineSegments.get(i);
            si.set(snew.get(i));
        }
    }

    public void overlapping_line_removal() {
        overlapping_line_removal(-10000.0);
    }

    /**
     * Divide the two line segments at the intersection of the two intersecting line segments. If there were two line segments that completely overlapped, both would remain without any processing.
     */
    public void intersect_divide() throws InterruptedException {
        int i_divide = 1;//1 if there is a split, 0 if not

        ArrayList<Boolean> k_flg = new ArrayList<>();//A flag that indicates that there is an effect of crossing.

        for (int i = 0; i < lineSegments.size(); i++) {
            k_flg.add(true);
        }

        while (i_divide != 0) {
            i_divide = 0;
            for (int i = 0; i < lineSegments.size(); i++) {
                if (k_flg.get(i)) {
                    k_flg.set(i, false);
                    for (int j = 0; j < lineSegments.size(); j++) {
                        if (i != j) {
                            if (k_flg.get(j)) {
                                int old_sousuu = lineSegments.size();
                                boolean itemp = intersect_divide(i, j); // Side effect
                                if (old_sousuu < lineSegments.size()) {
                                    for (int is = old_sousuu; is < lineSegments.size(); is++) {
                                        k_flg.add(true);
                                    }
                                }
                                if (itemp) {
                                    i_divide++;
                                    k_flg.set(i, true);
                                }
                            }
                        }
                    }
                }
            }

            if (Thread.interrupted()) throw new InterruptedException();
        }
    }

    /**
     * Divide the two line segments at the intersection of the two intersecting line segments. After splitting 1. Returns 0 if not done. From Orihime 2.002, the color of the line after splitting is also controlled (if there is an overlap, it will be unified and the color will be the one with the later number).
     */
    public boolean intersect_divide(int i, int j) {
        if (i == j) {
            return false;
        }

        LineSegment si = lineSegments.get(i);
        LineSegment sj = lineSegments.get(j);

        Point p1 = new Point(si.getA());
        Point p2 = new Point(si.getB());
        Point p3 = new Point(sj.getA());
        Point p4 = new Point(sj.getB());
        Point pk = new Point();

        double ixmax = Math.max(si.getAX(), si.getBX());
        double ixmin = Math.min(si.getAX(), si.getBX());
        double iymax = Math.max(si.getAY(), si.getBY());
        double iymin = Math.min(si.getAY(), si.getBY());

        double jxmax = Math.max(sj.getAX(), sj.getBX());
        double jxmin = Math.min(sj.getAX(), sj.getBX());
        double jymax = Math.max(sj.getAY(), sj.getBY());
        double jymin = Math.min(sj.getAY(), sj.getBY());

        if (ixmax + 0.5 < jxmin) {
            return false;
        }
        if (jxmax + 0.5 < ixmin) {
            return false;
        }
        if (iymax + 0.5 < jymin) {
            return false;
        }
        if (jymax + 0.5 < iymin) {
            return false;
        }

        LineSegment.Intersection intersect_decide = OritaCalc.determineLineSegmentIntersection(si, sj);
        switch (intersect_decide) {
            case INTERSECTS_1:
                pk.set(OritaCalc.findIntersection(si, sj));

                si.setA(p1);
                si.setB(pk);
                sj.setA(p3);
                sj.setB(pk);
                addLine(p2, pk, si.getColor());
                addLine(p4, pk, sj.getColor());
                return true;
            case INTERSECTS_TSHAPE_S1_VERTICAL_BAR_25:
            case INTERSECTS_TSHAPE_S1_VERTICAL_BAR_26:
                pk.set(OritaCalc.findIntersection(si, sj));

                sj.setA(p3);
                sj.setB(pk);
                addLine(p4, pk, sj.getColor());
                return true;

            case INTERSECTS_TSHAPE_S2_VERTICAL_BAR_27:
            case INTERSECTS_TSHAPE_S2_VERTICAL_BAR_28:
                pk.set(OritaCalc.findIntersection(si, sj));

                si.setA(p1);
                si.setB(pk);
                addLine(p2, pk, si.getColor());
                return true;

            case PARALLEL_EQUAL_31: //If the two line segments are exactly the same, do nothing.
                return false;
            case PARALLEL_START_OF_S1_CONTAINS_START_OF_S2_321: {//The endpoints of two line segments (p1 and p3) overlap at one point. si contains sj
                si.setA(p2);
                si.setB(p4);

                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                sj.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_START_OF_S2_CONTAINS_START_OF_S1_322: {//The endpoints of two line segments (p1 and p3) overlap at one point. sj contains si
                sj.setA(p2);
                sj.setB(p4);
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                si.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_START_OF_S1_CONTAINS_END_OF_S2_331: {//The endpoints of two line segments (p1 and p4) overlap at one point. si contains sj
                si.setA(p2);
                si.setB(p3);

                LineColor overlapping_col;
                overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                sj.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_END_OF_S2_CONTAINS_START_OF_S1_332: {//The endpoints of two line segments (p1 and p4) overlap at one point. sj contains si
                sj.setA(p2);
                sj.setB(p3);
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                si.setColor(overlapping_col);
                return true;
            }
            case PARALLEL_END_OF_S1_CONTAINS_START_OF_S2_341: {//The endpoints of two line segments (p2 and p3) overlap at one point. si contains sj
                si.setA(p1);
                si.setB(p4);
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                sj.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_START_OF_S2_CONTAINS_END_OF_S1_342: {//The endpoints of two line segments (p2 and p3) overlap at one point. sj contains si
                sj.setA(p1);
                sj.setB(p4);
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                si.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_END_OF_S1_CONTAINS_END_OF_S2_351: {//The endpoints of two line segments (p2 and p4) overlap at one point. si contains sj
                si.setA(p1);
                si.setB(p3);

                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                sj.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_END_OF_S2_CONTAINS_END_OF_S1_352: {//The endpoints of two line segments (p2 and p4) overlap at one point. sj contains si
                sj.setA(p1);
                sj.setB(p3);
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                si.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_S1_INCLUDES_S2_361: {//In order of p1-p3-p4-p2
                si.setA(p1);
                si.setB(p3);

                addLine(p2, p4, si.getColor());
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                sj.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_S1_INCLUDES_S2_362: {//In order of p1-p4-p3-p2
                si.setA(p1);
                si.setB(p4);

                addLine(p2, p3, si.getColor());

                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                sj.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_S2_INCLUDES_S1_363: {//In order of p3-p1-p2-p4
                sj.setA(p1);
                sj.setB(p3);

                addLine(p2, p4, sj.getColor());

                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                si.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_S2_INCLUDES_S1_364: {//In order of p3-p2-p1-p4
                sj.setA(p1);
                sj.setB(p4);

                addLine(p2, p3, sj.getColor());

                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                si.setColor(overlapping_col);

                return true;
            }
            case PARALLEL_S1_END_OVERLAPS_S2_START_371: {//In order of p1-p3-p2-p4
                si.setA(p1);
                si.setB(p3);

                sj.setA(p2);
                sj.setB(p4);

                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                addLine(p2, p3, overlapping_col);
                return true;
            }
            case PARALLEL_S1_END_OVERLAPS_S2_END_372: {//In order of p1-p4-p2-p3
                si.setA(p1);
                si.setB(p4);

                sj.setA(p3);
                sj.setB(p2);

                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                addLine(p2, p4, overlapping_col);
                return true;
            }
            case PARALLEL_S1_START_OVERLAPS_S2_END_373: {//In order of p3-p1-p4-p2
                sj.setA(p1);
                sj.setB(p3);
                si.setA(p2);
                si.setB(p4);
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                addLine(p1, p4, overlapping_col);
                return true;
            }
            case PARALLEL_S1_START_OVERLAPS_S2_START_374: {//In order of p4-p1-p3-p2
                sj.setA(p1);
                sj.setB(p4);
                si.setA(p3);
                si.setB(p2);
                LineColor overlapping_col = si.getColor();
                if (i < j) {
                    overlapping_col = sj.getColor();
                }
                addLine(p1, p3, overlapping_col);
                return true;
            }
            default:
                return false;
        }
    }

    /**
     * Add line segment
     */
    private void addLine(Point pi, Point pj, LineColor i_c) {
        LineSegment s = new LineSegment();
        s.set(pi, pj, i_c);

        lineSegments.add(s);
    }
}
