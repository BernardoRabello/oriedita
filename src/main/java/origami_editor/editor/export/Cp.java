package origami_editor.editor.export;

import origami.crease_pattern.element.LineColor;
import origami.crease_pattern.element.LineSegment;
import origami_editor.editor.Save;
import origami_editor.record.Memo;

import java.util.StringTokenizer;

public class Cp {
    public static Memo exportFile(Save save) {
        Memo mem = new Memo();
        for (LineSegment s :
                save.getLineSegments()) {
            int color = s.getColor().getNumber() + 1;

            mem.addLine(String.format("%d %s %s %s %s", color, s.getAX(), s.getAY(), s.getBX(), s.getBY()));
        }

        return mem;
    }

    public static Save importFile(Memo mem) {
        Save save = new Save();

        double d1, d2, d3, d4;

        //オリヒメ　0.Contour, 1.Mountain, 2.Valley　、ORIPA 1.Contour, 2.Mountain, 3.Valley

        for (int ig = 1; ig <= mem.getLineCount(); ig++) {
            if (mem.getLine(ig).length() != 0) {
                StringTokenizer tk = new StringTokenizer(mem.getLine(ig), " ");
                String str = tk.nextToken();
                LineColor col;
                switch (str) {
                    case "1":
                        col = LineColor.BLACK_0;
                        break;
                    case "2":
                        col = LineColor.RED_1;
                        break;
                    case "3":
                        col = LineColor.BLUE_2;
                        break;
                    default:
                        continue;
                }
                d1 = Double.parseDouble(tk.nextToken());
                d2 = Double.parseDouble(tk.nextToken());
                d3 = Double.parseDouble(tk.nextToken());
                d4 = Double.parseDouble(tk.nextToken());


                LineSegment s = new LineSegment();
                s.set(d1, d2, d3, d4, col);

                save.addLineSegment(s.clone());
            }
        }

        return save;
    }
}
