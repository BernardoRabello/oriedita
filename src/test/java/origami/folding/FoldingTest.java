package origami.folding;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import origami.crease_pattern.FoldingException;
import origami.crease_pattern.LineSegmentSet;
import origami.crease_pattern.element.Point;
import origami_editor.editor.Save;
import origami_editor.editor.component.BulletinBoard;
import origami_editor.editor.export.Cp;

import java.io.IOException;

@ExtendWith({SnapshotExtension.class})
public class FoldingTest {
    private Expect expect;

    @Test
    public void testFoldBirdbase() throws IOException {
        Save save = Cp.importFile(getClass().getClassLoader().getResourceAsStream("birdbase.cp"));

        FoldedFigure foldedFigure = new FoldedFigure(new BulletinBoard());

        LineSegmentSet lineSegmentSet = new LineSegmentSet();

        lineSegmentSet.setSave(save);

        try {
            foldedFigure.estimationOrder = FoldedFigure.EstimationOrder.ORDER_5;
            foldedFigure.folding_estimated(lineSegmentSet, new Point());

            expect.serializer("json").toMatchSnapshot(foldedFigure.cp_worker3.get());
        } catch (InterruptedException | FoldingException e) {
            e.printStackTrace();
            Assertions.fail();
        }
    }
}
