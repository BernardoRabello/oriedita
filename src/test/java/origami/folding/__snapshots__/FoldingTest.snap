origami.folding.FoldingTest.testSimpleCreasePattern=[
  {
    "numFaces_temp": 103,
    "numPoints": 5,
    "numLines": 8,
    "numFaces": 4,
    "points": [
      "0.0,0.0",
      "-200.0,-200.0",
      "-200.0,3.7665327494283095E-30",
      "-200.0,82.842712474619",
      "-117.157287525381,-9.021118369802149E-17",
      "-282.84271247461896,1.4210854715202004E-14"
    ],
    "lines": [
      {
        "begin": 0,
        "end": 0,
        "color": "BLACK_0"
      },
      {
        "begin": 1,
        "end": 2,
        "color": "BLUE_2"
      },
      {
        "begin": 2,
        "end": 3,
        "color": "BLUE_2"
      },
      {
        "begin": 2,
        "end": 4,
        "color": "BLUE_2"
      },
      {
        "begin": 2,
        "end": 5,
        "color": "BLUE_2"
      },
      {
        "begin": 1,
        "end": 5,
        "color": "RED_1"
      },
      {
        "begin": 3,
        "end": 5,
        "color": "RED_1"
      },
      {
        "begin": 1,
        "end": 4,
        "color": "RED_1"
      },
      {
        "begin": 3,
        "end": 4,
        "color": "RED_1"
      }
    ],
    "faces": [
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0,
          1,
          2,
          4
        ],
        "numPoints": 3,
        "icol": 0
      },
      {
        "pointIdList": [
          0,
          1,
          5,
          2
        ],
        "numPoints": 3,
        "icol": 0
      },
      {
        "pointIdList": [
          0,
          2,
          3,
          4
        ],
        "numPoints": 3,
        "icol": 0
      },
      {
        "pointIdList": [
          0,
          2,
          5,
          3
        ],
        "numPoints": 3,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      },
      {
        "pointIdList": [
          0
        ],
        "numPoints": 0,
        "icol": 0
      }
    ]
  }
]